2단계 - 로또

yangseungin opened the pull request:
안녕하세요! 아직 미션을 다 완성하지 못하였는데 궁금한점이 있어서 PR을 생성하게되었습니다.
자세한 요구사항 조건이 없어서 일반적으로 알고있는 요구사항으로 기능 요구사항을 뽑아봤습니다.
변수명을 뽑으면서 알고있는  영어단어가 짧음을 또 한번 느끼고있네요.
요구사항 정리를 할때 어떠한 기능들 단위로 클래스를 뽑을지에 시간 투자를 많이하였는데도 불구하고 구현하다 보니 계속 바뀌게 되네요
PR작성하고나서 보니 랜덤에 대한 테스트도 작성하기 힘들게 작성한것같아서 수정해보겠습니다

Q1.
TDD사이클을 연습하면서 현재 가장 작은 단위의 기능이 될거로 예상되는 LottoNumber, 그다음 Lotto 그리고 여러장이모인 LottoTickets 그리고 당첨갯수와 상금을 계산하는 Rank까지 구현하였는데요 
Lotto에서는 번호가 중복인경우에대한 validate를 진행하였는데 "랜덤 로또를 생성한다"는 책임을 LottoTickets에 우선 넣어놨는데 Lotto로 들어가도 이상하지 않을것 같기도하고해서 이런 부분이 조금 애매하네요

yangseungin committed:
refactor: 주생성자를 활용하도록 변경

yangseungin committed:
docs: 사전지식을 바탕으로 lotto 요구사항을 정리한다

yangseungin committed:
feat: 로또번호를 생성한다.

yangseungin committed:
feat: 로또를 생성한다.

yangseungin committed:
feat: 로또티켓을 생성한다

yangseungin committed:
feat: Rank를 구현한다
- 매칭갯수에 따른 상금계산을 위임한다

yangseungin committed:
docs: 완료한 Rank 요구사항 체크

javajigi commented on src/main/java/lotto/Lotto.java:
```diff
@@ -0,0 +1,71 @@
+        validateSize(numbers);
+        validateDuplicate(numbers);
+        this.numbers = createLottoNumbers(numbers);
+    }
```
```suggestion
    public Lotto(List<Integer> numbers) {
        this(createLottoNumbers(numbers));
    }
    
    public Lotto(List<LottoNumber> numbers) {
        validateSize(numbers);
        validateDuplicate(numbers);
        this.numbers = numbers;
    }
```

위와 같이 주 생성자를 추가하고 부 생성자가 주 생성자를 호출하도록 구현하면 어떨까?

yangseungin replied:
지속적으로 리뷰해주시는 내용인데 이부분이 바로 잘안되는듯하네요 신경써서 진행해보겠습니다

javajigi commented on src/main/java/lotto/Lotto.java:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    private void validateDuplicate(List<Integer> numbers) {
+        Set<Integer> uniqueNumbers = new HashSet<>(numbers);
```
중복을 허용하지 않는다면 인스턴스 변수를 List가 아니라 List<LottoNumber>으로 구현하는 것은 어떨까?

yangseungin replied:
Set을 고려하지 않은건 아닌데 출력 예시에서 순서를 보장하는듯한 느낌이라 List로구현하긴 했었습니다.

javajigi replied:
String을 split해 정렬 후 List로 생성하는데요.
Set으로 구현한 후 값을 꺼낼 때 정렬해서 반환해도 되지 않을까요?

javajigi commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,71 @@
+    public int countMatchNumber(Lotto other) {
+        int count = 0;
+        for (LottoNumber number : numbers) {
+            if (other.contains(number)) {
```
👍

javajigi commented on src/main/java/lotto/Lotto.java:
```diff
@@ -0,0 +1,71 @@
+                return true;
+            }
+        }
+        return false;
```
```suggestion
        return this.numbers.contains(number);
```
콜렉션의 contains()를 활용해 구현하면 어떨까?

javajigi commented on src/main/java/lotto/domain/LottoNumber.java:
```diff
@@ -0,0 +1,46 @@
+
+import java.util.Objects;
+
+public class LottoNumber {
```
LottoNumber 값으로 추상화한 것 💯

javajigi commented on src/main/java/lotto/LottoNumber.java:
```diff
@@ -0,0 +1,46 @@
+
+    public LottoNumber(String value) {
+        this(Integer.parseInt(value));
+    }
```
```suggestion
    public LottoNumber(String value) {
        this(Integer.parseInt(value));
    }
    
    public LottoNumber(int value) {
        validate(value);
        this.value = value;
    }
```

주 생성자를 마지막에 구현하는 것이 컨벤션

javajigi commented on src/main/java/lotto/LottoTickets.java:
```diff
@@ -0,0 +1,51 @@
+        this.lottos = lottos;
+    }
+
+    public static LottoTickets create(int purchaseAmount) {
```
purchaseAmount도 원시 값인데 포장해 보면 어떨까?
포장한 후에 이 객체의 역할에 대해 고민해 본다.

yangseungin replied:
구매 금액으로 사용했었는데변수명을 잘못지었네요
원시값으로 포장하면 금액을 입력하였을때 음수에 대한 validation이나 1000원단위로 입력해야한다는 조건등을 처리할 수 있을 것 같아요

javajigi commented on src/main/java/lotto/LottoTickets.java:
```diff
@@ -0,0 +1,51 @@
+    }
+
+    private static Lotto createRandomLotto() {
+        //TODO 많은 일을 하고 있음
```
공감함
메서드를 분리하거나 새로운 객체로 분리하는 것은 어떨까?

javajigi commented on src/test/java/lotto/LottoTest.java:
```diff
@@ -0,0 +1,72 @@
+    void 로또_번호_6개로_생성() {
+        List<Integer> numbers = Arrays.asList(1, 2, 3, 4, 5, 6);
+
+        Lotto lotto = new Lotto(numbers);
```
```suggestion
        Lotto lotto = new Lotto(1, 2, 3, 4, 5, 6);
```
가변 인자를 받는 생성자도 추가하면 Lotto 생성이 더 편리해 지지 않을까?

javajigi reviewed (changes requested):
자동차 경주 미션에서 지켰던 원칙을 지키면서 구현해서 일까요?
전체적인 객체 설계, 포장, 테스트 코드 구현 등이 작은 단위로 깔끔하게 구현 잘 했네요. 💯 
그럼에도 불구하고 몇 가지 개선했으면 하는 부분이 있어 피드백 남겼어요.

Q1.
TDD사이클을 연습하면서 현재 가장 작은 단위의 기능이 될거로 예상되는 LottoNumber, 그다음 Lotto 그리고 여러장이모인 LottoTickets 그리고 당첨갯수와 상금을 계산하는 Rank까지 구현하였는데요
Lotto에서는 번호가 중복인경우에대한 validate를 진행하였는데 "랜덤 로또를 생성한다"는 책임을 LottoTickets에 우선 넣어놨는데 Lotto로 들어가도 이상하지 않을것 같기도하고해서 이런 부분이 조금 애매하네요

A1.
이 부분 또한 정답이 있는 것은 아닌 영역인데요.
저는 LottoTickets와 Lotto 모두 랜덤 로또를 생성 책임을 부여하기에 적합한 곳은 아니지 않나라는 생각을 해봅니다.
객체의 역할을 어떻게 부여하느냐의 차이인 만큼 승인님이 결정하고 구현하고 리뷰 요청해보세요.

yangseungin committed:
refactor: 주생성자, 정적팩토리메서드추가 및 정렬 메서드 분리

yangseungin committed:
chore: 컨벤션 정리

yangseungin committed:
refactor: 부생성자 추가

yangseungin committed:
feat: 당첨결과를 구현한다

yangseungin committed:
chore: 패키지 구조 변경

yangseungin committed:
feat: InputView, OutputView 구현

yangseungin committed:
feat: PurchaseAmount 원시값 포장

yangseungin committed:
refactor: 셀프 피드백 개선
- Winningresult: 주생성자로 필드 를 직접 초기화하도록 개선 및 원시값 포장 객체 사용하도록 수정
- ResultView: InputView와 동일하게 static

yangseungin committed:
feat: 로또번호 Generator 분리

yangseungin committed:
docs: 요구사항 수정

yangseungin committed:
docs: 피드백 반영 후 요구사항 수정

yangseungin commented:
LottoTickets의 createRandomLotto가 1~45 숫자 리스트도 만들고 섞어서 6개를 선택하고 로또까지 생성하는 많은 일을 하고있어서 LottoNumberGenerator로 번호생성관련 역할을 위임해봤습니다.

javajigi commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,86 @@
+
+    private final List<LottoNumber> numbers;
+
+    public Lotto(int... intNumbers) {
```
👍

javajigi commented on src/main/java/lotto/domain/LottoNumberGenerator.java:
```diff
@@ -0,0 +1,33 @@
+        return numbers;
+    }
+
+    private static List<Integer> selectNumbers(List<Integer> numbers) {
+        List<Integer> selectedNumbers = new ArrayList<>();
+        for (int i = 0; i < LOTTO_NUMBER_COUNT; i++) {
+            selectedNumbers.add(numbers.get(i));
+        }
+        return selectedNumbers;
+    }
```
```suggestion
    private static List<Integer> selectNumbers(List<Integer> numbers) {
        return new ArrayList(numbers.sublist(0, 6));
    }
```
콜렉션의 sublist 사용해 구현하면 어떨까?

yangseungin replied:
콜렉션의 사용을 최대한 지양하고 구현했었는데 stream api를 사용을 지양하자고했지 콜렉션은 상관이없었네요

javajigi commented on src/main/java/lotto/domain/LottoTickets.java:
```diff
@@ -0,0 +1,31 @@
+import java.util.List;
+
+public class LottoTickets {
+    private static final int LOTTO_PRICE = 1000;
```
```suggestion
    private static final int LOTTO_PRICE = 1_000;
```

javajigi commented on src/main/java/lotto/domain/LottoTickets.java:
```diff
@@ -0,0 +1,31 @@
+    }
+
+    public static LottoTickets create(PurchaseAmount purchaseAmount) {
+        int count = purchaseAmount.getAmount() / LOTTO_PRICE;
```
PurchaseAmount이 몇 장의 로또를 구매할 수 있는지에 대한 책임을 부여하면 어떨까?
객체의 값을 꺼내지 말고 메시지를 보내라

javajigi commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -0,0 +1,86 @@
+        return result;
+    }
+
+    public int countMatchNumber(Lotto other) {
```
```suggestion
    public Rank match(Lotto other) {
```
이미 Rank도 enum으로 구현한 상태라면 Lotto가 몇 등인지 여부를 판단하는 것은 어떨까?

yangseungin replied:
로또에게 다른로또를 가지고 굳이 숫자가 몇개 맞는지를 확인할 필요없이 몇등인지 묻는방식이 좋은것 같네요

javajigi commented on src/main/java/lotto/domain/WinningResult.java:
```diff
@@ -0,0 +1,47 @@
+import java.util.Map;
+
+public class WinningResult {
+    private final Map<Rank, Integer> result;
```
현재와 같이 구현도 👍 
추가로 객체를 분리한다면 Map의 key/value(Rank, Integer)를 새로운 객체로 분리하는 것이 로직 구현에 의미가 있을까?

yangseungin replied:
Map<Rank, Integer>를 새로운 객체로 분리하면 특정등수의갯수?라는 의미를 갖게될거같은데 지금도 충분한 느낌이긴하네요

javajigi commented on src/main/java/lotto/domain/WinningResult.java:
```diff
@@ -0,0 +1,47 @@
+    }
+
+    public double calculateProfitRate(PurchaseAmount purchaseAmount) {
+        return (double) getTotalWinningAmount() / purchaseAmount.getAmount();
```
PurchaseAmount을 Money라는 이름으로 바꾸고 입력하는 돈, 수익률 계산 등 돈과 관련한 모든 로직을 Money와 같은 객체를 추가해 구현해 보는 것은 어떨까?
지금도 객체에 메시지를 보내기보다 값을 꺼내 로직을 구현하고 있음

yangseungin replied:
기존의 PurchaseAmount는 Money의 성격이 강한듯해서  Money로 변경하고 PurchaseAmount는 몇 장의 로또를 구매할 수 있는지의 책임을 가지도록 수정해보겠습니다.

javajigi reviewed (approved):
피드백 반영하느라 수고했어요. 👍 
로또 생성하는 부분에 대해 새로운 객체로 분리 잘 했네요.
아직도 원시값을 포장한 후 값을 꺼내 로직을 구현하는 절차지향적인 사고를 하는 부분이 보이네요.
상태 값을 가지는, 도메인 로직을 구현하는 객체를 수동적인 존재가 아니라 능동적이고, 주체적인 존재로 바라보기 위해 의식적인 연습을 해보면 좋겠습니다.
피드백 남겼는데요. 3단계 진행하면서 함께 반영해 보시죠.
